PowerPointLabs/PowerPointLabs/Utils/ClipboardUtil.cs

```diff
@@ -22,29 +23,209 @@ public static ShapeRange PasteShapesFromClipboard(PowerPointSlide slide)
+            if (pastedShapes.Count >= 1 && finalSlideShapesCount == initialSlideShapesCount)
+            {
+                pastedShapes.Cut();
+                if (!isPasteSpecial)
                 {
-                    pastedShapes.Cut();
                     pastedShapes = slide.Shapes.Paste();
                 }
+                else
+                {
+                    pastedShapes = slide.Shapes.PasteSpecial(pasteType);
+                }
+            }
+
+            return pastedShapes;
+        }
 
-                return pastedShapes;
+        private static SlideRange TryPastingAsSlide(PowerPointPresentation pres)
+        {
+            try
+            {
+                // try pasting as slide
+                SlideRange slides = pres.PasteSlide();
+                return (slides.Count >= 1) ? slides : null;
             }
             catch (COMException e)
             {
-                // May be thrown if there is placeholder shape in clipboard
-                Logger.LogException(e, "PasteShapeFromClipboard");
+                // May be thrown if clipboard is not a slide
+                Logger.LogException(e, "RestoreClipboardAfterAction: pasting as slide");
                 return null;
             }
         }
 
-        #endregion
+        private static ShapeRange TryPastingAsText(PowerPointSlide slide)
+        {
+            try
+            {
+                // try pasting as text
+                return PasteWithCorrectSlideCheck(slide, true, PpPasteDataType.ppPasteText);
+            }
+            catch (COMException e)
+            {
+                // May be thrown if clipboard is not text
+                Logger.LogException(e, "RestoreClipboardAfterAction: pasting as text");
+                return null;
+            }
+        }
+
+        private static ShapeRange TryPastingAsShape(PowerPointSlide slide) 
+        {
+            try
+            {
+                // try pasting as shape
+                return PasteWithCorrectSlideCheck(slide, true, PpPasteDataType.ppPasteShape);
+            }
+            catch (COMException e)
+            {
+                // May be thrown if clipboard is not a shape
+                Logger.LogException(e, "TryPastingAsShape");
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Pastes clipboard content into new temp slide using the DocumentWindow's View.Paste()
+        /// Though this paste will work for most clipboard objects (even web pictures), it will change the undo history.
+        /// </summary>
+        private static Shape TryPastingOntoView(PowerPointPresentation pres, PowerPointSlide tempSlide, PowerPointSlide origSlide)
+        {
+            try
+            {
+                // Utilises deprecated Globals class as ClipboardUtil does not utilise ActionFramework
+                DocumentWindow workingWindow = Globals.ThisAddIn.Application.ActiveWindow;
```

Comment:
@IanTeo @Skaty  I'm unsure of how I should use ActionFramework with ClipboardUtil, should I make an extension instead for this Utils class? (Eg. Make a BaseUtil class and an extension for BaseUtil that uses the ActionFramework)

Reply:
i'll just use Globals so that we don't have to extend the ActionFramework unnecessarily.

Reply:
@IanTeo @Skaty I shall stick to using Globals here so that we don't have to extend the ActionFramework unnecessarily?